### PR TITLE
Add support for JSON string mods

### PR DESCRIPTION
**Pull request by PowerBall253**
Adds support for loading string JSONs into the .blang files, using idCrypt to decrypt/encrypt the .blang, BlangParser to parse it, and NewtonSoft Json to parse the JSON.
The JSON files must look like this:

```json
{
    "strings":[
        {
            "name":"#str_decl_rune_targetstrike_title_7_GHOST54171",
            "text":"Hello world"
        },
        {
            "name":"#str_decl_rune_targetstrike_desc_GHOST54172",
            "text":"Hi!"
        },
        {
            "name":"#str_custom_string",
            "text":"This last one doesn't replace any existing strings!\nIsn't that wonderful?"
        }
    ]
}
```

The path for the JSON files in a mod is `(resource name)/EternalMod/strings/(language).json`.

This makes possible to have various mods editing strings, without overwriting each other.

### CODE REVIEW

**Comment by ZwipZwapZapony** [comment]
I can confirm that it works. I've tried it with two zip mods and a loose-file mod that all changed/added their own strings, and all three mods' strings were visible in-game.

That said, I've only tried editing the English language file, not any others, and I haven't done much code review either.

File: EternalModLoader/BlangJson.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+{
+    /// <summary>
+    /// BlangJson class for JSON deserialization
+    /// </summary>
+    public class BlangJson
+    {
+        /// <summary>
+        /// List of StringObjects for JSON deserialization
+        /// </summary>
+        public IList<StringObject> strings { get; set; }
```
Comment (dcealopez) [inline EternalModLoader/BlangJson.cs:13]: Please follow C# coding conventions. See: https://docs.microsoft.com/es-es/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

Please use PascalCase for properties.

If the JSON must be serialized, tell the serializer to use camelCase instead.

File: EternalModLoader/BlangJson.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    
+    /// <summary>
+    /// StringObject class used by BlangJson class
+    /// </summary>
+    public class StringObject
+    {
+        /// <summary>
+        /// String identifier
+        /// </summary>
+        public string name { get; set; }
```
Comment (dcealopez) [inline EternalModLoader/BlangJson.cs:24]: Please follow C# coding conventions. See: https://docs.microsoft.com/es-es/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

Please use PascalCase for properties.

If the JSON must be serialized, tell the serializer to use camelCase instead.

File: EternalModLoader/BlangJson.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        
+        /// <summary>
+        /// String text
+        /// </summary>
+        public string text { get; set; }
```
Comment (dcealopez) [inline EternalModLoader/BlangJson.cs:29]: Please follow C# coding conventions. See: https://docs.microsoft.com/es-es/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

Please use PascalCase for properties.

If the JSON must be serialized, tell the serializer to use camelCase instead.

File: EternalModLoader/BlangParser.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace EternalModLoader
+{
+    /// <summary>
+    /// BlangFile class
+    /// </summary>
+    public class BlangFile
```
Comment (dcealopez) [inline EternalModLoader/BlangParser.cs:10]: This shouldn't be here. You should reference the BlangParser library that you can find here: https://github.com/dcealopez/BlangParser

This way, if I implement changes to the BlangParser library, they will be automatically traslated to the mod loader.

File: EternalModLoader/BlangString.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+namespace EternalModLoader
+{
+    /// <summary>
+    /// BlangString class
+    /// </summary>
+    public class BlangString
```
Comment (dcealopez) [inline EternalModLoader/BlangString.cs:6]: This shouldn't be here. You should reference the BlangParser library that you can find here: https://github.com/dcealopez/BlangParser

This way, if I implement changes to the BlangParser library, they will be automatically traslated to the mod loader.

File: EternalModLoader/EternalModLoader.cs [hunk 6]
```diff
@@ -249,19 +250,112 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
...
+                        if (res != 0)
+                        {
+                            continue;
+                        }
+
+                        BlangFile blangFile;
+                        using (var blangMemoryStream = new MemoryStream(blangFileBytes))
+                        {
+                            blangFile = BlangFile.Parse(blangMemoryStream);
+
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:320]: Unnecessary empty line.

File: EternalModLoader/BlangString.cs (hunk 5 above)
Comment (PowerBall253) [reply EternalModLoader/BlangString.cs:6]: There is a problem with this though, the methods are modified.

File: EternalModLoader/EternalModLoader.cs [hunk 7]
```diff
@@ -249,19 +250,112 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
...
                     long sizeDiff = mod.FileBytes.Length - size;
 
+                    // If the mod is a blang file json, modify the .blang file
+                    if (mod.isBlangJson)
+                    {
+                        BlangJson blangJson;
+                        try
+                        {
+                            blangJson = JsonConvert.DeserializeObject<BlangJson>(Encoding.UTF8.GetString(mod.FileBytes));
+                            if (blangJson == null || blangJson.strings.Count == 0)
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:289]: Always use braces, even for 1 line conditions. Please follow the C# coding convention: https://docs.microsoft.com/es-es/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

File: EternalModLoader/EternalModLoader.cs [hunk 8]
```diff
@@ -249,19 +250,112 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
...
+                                throw new Exception();
+                            foreach (var blangJsonString in blangJson.strings)
+                            {
+                                if (blangJsonString == null || blangJsonString.name == null || blangJsonString.text == null)
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:293]: Always use braces, even for 1 line conditions. Please follow the C# coding convention: https://docs.microsoft.com/es-es/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

File: EternalModLoader/EternalModLoader.cs [hunk 9]
```diff
@@ -249,19 +250,112 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
...
+                                if (blangJsonString.name.Equals(blangString.Identifier))
+                                {
+                                    stringFound = true;
+                                    blangString.Text = blangJsonString.text;
+                                    Console.WriteLine($"\tReplaced {blangString.Identifier} in {mod.Name}");
+                                    break;
+                                }
+                            }
+
+                            if (stringFound)
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:338]: Always use braces, even for 1 line conditions. Please follow the C# coding convention: https://docs.microsoft.com/es-es/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

File: EternalModLoader/EternalModLoader.cs [hunk 10]
```diff
@@ -320,7 +414,9 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
                         }
                     }
 
-                    Console.WriteLine(string.Format("\tReplaced {0}", mod.Name));
+                    if (!mod.isBlangJson)
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:417]: Always use braces, even for 1 line conditions. Please follow the C# coding convention: https://docs.microsoft.com/es-es/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

File: EternalModLoader/EternalModLoader.cs [hunk 11]
```diff
@@ -665,11 +761,29 @@ public static int Main(string[] args)
                             var stream = zipArchive.GetEntry(modFileName).Open();
 
                             using (var memoryStream = new MemoryStream())
-                            {
+                            {
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:764]: Please remove the trailing whitespace.

File: EternalModLoader/Mod.cs [hunk 12]
```diff
@@ -14,6 +14,11 @@ public class Mod
         /// Mod file bytes
         /// </summary>
         public byte[] FileBytes;
+        
+        /// <summary>
+        /// True if the mod file is a .blang json
+        /// </summary>
+        public bool isBlangJson;
```
Comment (dcealopez) [inline EternalModLoader/Mod.cs:21]: Please follow C# coding conventions. See: https://docs.microsoft.com/es-es/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

Please use PascalCase for properties.

File: EternalModLoader/BlangDecrypt.cs [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+                        using (CryptoStream csEncrypt = new CryptoStream(msEncrypt, encryptor, CryptoStreamMode.Write))
+                        {
+                            csEncrypt.Write(pbInput, 0, pbInput.Length);
+                            csEncrypt.FlushFinalBlock();
+                            
+                            return msEncrypt.ToArray();
+                        }
+                    }
+                }
+
```
Comment (dcealopez) [inline EternalModLoader/BlangDecrypt.cs:213]: Unnecessary empty line.

File: EternalModLoader/EternalModLoader.cs [hunk 14]
```diff
@@ -249,19 +250,112 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
...
                     {
-                        resourceInfo.ModListNew.Add(mod);
-                        continue;
+                        var modName = mod.Name;
+                        var modFilePathParts = modName.Split('/');
+                        var name = modName.Remove(0, modFilePathParts[0].Length + 1);
+                        mod.Name = name.Substring(0, name.Length - 4) + "blang";
+                        chunk = GetChunk(mod.Name, resourceInfo);
+
+                        if (chunk == null)
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:263]: Always use braces, even for 1 line conditions. Please follow the C# coding convention: https://docs.microsoft.com/es-es/dotnet/csharp/programming-guide/inside-a-program/coding-conventions

File: EternalModLoader/EternalModLoader.cs [hunk 15]
```diff
@@ -249,19 +250,112 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
...
+                            }
+
+                            if (stringFound)
+                                continue;
+                            
+                            blangFile.Strings.Add(new BlangString()
+                            {
+                                Identifier = blangJsonString.name,
+                                Text = blangJsonString.text,
+                            });
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:345]: Please separate this line with the next one with another line break.

File: EternalModLoader/EternalModLoader.cs [hunk 16]
```diff
@@ -249,19 +250,112 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
...
+                            Console.WriteLine($"\tAdded {blangJsonString.name} in {mod.Name}");
+                        }
+
+                        byte[] cryptDataBuffer = blangFile.WriteToBytes();
+                        res = BlangCrypt.IdCrypt(ref cryptDataBuffer, $"strings/{Path.GetFileName(mod.Name)}", false);
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:350]: Please separate this line with the next one with another line break.

File: EternalModLoader/BlangDecrypt.cs [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+        /// Encrypts or decrypts a blang file
+        /// </summary>
+        /// <param name="fileData">byte array contaning a blang file bytes</param>
+        /// <param name="internalPath">blang file's internal path</param>
+        /// <param name="decrypt">bool indicating if we wanna encrypt or decrypt</param>
+        /// <returns>int indicating success or failure</returns>
+        public static int IdCrypt(ref byte[] fileData, string internalPath, bool decrypt)
+        {
+            string keyDeriveStatic = "swapTeam\n";
+
```
Comment (dcealopez) [inline EternalModLoader/BlangDecrypt.cs:24]: Please join the previous and next lines together and separate them from the condition with an extra line break.

File: EternalModLoader/BlangDecrypt.cs [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+                {
+                    rng.GetBytes(fileSalt);
+                }
+            }
+
+            byte[] keyDeriveStaticBytes = new byte[0xA];
+            Array.Copy(Encoding.ASCII.GetBytes(keyDeriveStatic), keyDeriveStaticBytes, 0xA - 1);
+            keyDeriveStaticBytes[0xA - 1] = (byte)'\0';
+            
+            byte[] encKey = null;
```
Comment (dcealopez) [inline EternalModLoader/BlangDecrypt.cs:42]: Please separate this line from the next one with an extra line break.

File: EternalModLoader/BlangDecrypt.cs [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+            try
+            {
+                encKey = HashData(fileSalt, keyDeriveStaticBytes, Encoding.ASCII.GetBytes(internalPath), null);
+            }
+            catch
+            {
+                return 1;
+            }
+
+            byte[] fileIV = new byte[0x10];
```
Comment (dcealopez) [inline EternalModLoader/BlangDecrypt.cs:52]: Please separate this line from the next one with an extra line break.

File: EternalModLoader/BlangDecrypt.cs [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+            else
+            {
+                using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
+                {
+                    rng.GetBytes(fileIV);
+                }
+            }
+
+            byte[] fileText;
+            byte[] hmac = new byte[0x20];
```
Comment (dcealopez) [inline EternalModLoader/BlangDecrypt.cs:66]: Please separate this line from the next one with an extra line break.

File: EternalModLoader/BlangDecrypt.cs [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+                {
+                    return 1;
+                }
+            }
+            else
+            {
+                fileText = fileData;
+            }
+
+            byte[] cryptedText = null;
```
Comment (dcealopez) [inline EternalModLoader/BlangDecrypt.cs:94]: Please separate this line from the next one with an extra line break.

File: EternalModLoader/BlangDecrypt.cs [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+            try
+            {
+                cryptedText = CryptData(decrypt, fileText, encKey.Take(0x10).ToArray(), fileIV);
+            }
+            catch
+            {
+                return 1;
+            }
+
+            var cryptMemoryStream = new MemoryStream();
```
Comment (dcealopez) [inline EternalModLoader/BlangDecrypt.cs:104]: Please separate this line from the next one with an extra line break.

File: EternalModLoader/BlangDecrypt.cs [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+    internal class BlangCrypt
+    {
+        /// <summary>
+        /// Encrypts or decrypts a blang file
+        /// </summary>
+        /// <param name="fileData">byte array contaning a blang file bytes</param>
+        /// <param name="internalPath">blang file's internal path</param>
+        /// <param name="decrypt">bool indicating if we wanna encrypt or decrypt</param>
+        /// <returns>int indicating success or failure</returns>
+        public static int IdCrypt(ref byte[] fileData, string internalPath, bool decrypt)
```
Comment (dcealopez) [inline EternalModLoader/BlangDecrypt.cs:21]: Please add comments to this function explaining what's going on here, so it's more clear.

**Review by dcealopez — CHANGES_REQUESTED** [review]
Please check the comments added to the submitted code.

File: EternalModLoader/BlangDecrypt.cs [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+            }
+            else
+            {
+                using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
+                {
+                    rng.GetBytes(fileIV);
+                }
+            }
+
+            //Get plain text for AES
```
Comment (dcealopez) [inline EternalModLoader/BlangDecrypt.cs:70]: Add a space before the comment itself (after the //)

File: EternalModLoader/EternalModLoader.cs [hunk 25]
```diff
@@ -249,19 +252,124 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
...
+                            {
+                                Identifier = blangJsonString.Name,
+                                Text = blangJsonString.Text,
+                            });
+                            
+                            Console.WriteLine($"\tAdded {blangJsonString.Name} in {mod.Name}");
+                        }
+
+                        byte[] cryptDataBuffer = blangFile.WriteToStream().ToArray();
+
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:363]: Remove this empty line.

File: EternalModLoader/EternalModLoader.cs [hunk 26]
```diff
@@ -249,19 +252,124 @@ public static void ReplaceChunks(MemoryStream memoryStream, ResourceInfo resourc
...
+                                Text = blangJsonString.Text,
+                            });
+                            
+                            Console.WriteLine($"\tAdded {blangJsonString.Name} in {mod.Name}");
+                        }
+
+                        byte[] cryptDataBuffer = blangFile.WriteToStream().ToArray();
+                        
+                        res = BlangCrypt.IdCrypt(ref cryptDataBuffer, $"strings/{Path.GetFileName(mod.Name)}", false);
+                        if (res != 0)
```
Comment (dcealopez) [inline EternalModLoader/EternalModLoader.cs:365]: Add a line break before the if condition to separate the statement before it.

**Review by dcealopez — CHANGES_REQUESTED** [review]
Please review the latest comments.

**Review by dcealopez — APPROVED** [review]
Cool.

**Pull request merged**